calibre: build it from git source instead of tarball

wrvsrx opened the pull request:
## Description of changes

We used to build calibre using its official tarball. However, the tarball contains lots of generated code such as javascript generated from python, python generated from pyqt... It also contained generated man pages.

The main difficulty to build calibre from git source is that it downloads six packages from internet, some of them is irreproducible. This commit patches that code in the following principles:

1. The fewer changed the better.
2. If a downloaded package is not pinned to a rev, pin it to a latest rev.
3. If a downloaded package can't be pinned to a rev (such as downloading from `https://code.calibre-ebook.com/ua-popularity`), store it in nixpkgs.
4. Allow passing downloaded packages using environment variable so user can override them when needed

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wrvsrx committed:
calibre: build it from git source instead of tarball

We used to build calibre using its official tarball. However, the
tarball contains lots of generated code such as javascript generated
from python, python generated from pyqt... It also contained generated
man pages.

The main difficulty to build calibre from git source is that it
downloads six packages from internet, some of them is irreproducible. This
commit patches that code in the following principles:

1. The fewer changed the better.
2. If a downloaded package is not pinned to a rev, pin it to a latest rev.
3. If a downloaded package can't be pinned to a rev (such as downloading
   from `https://code.calibre-ebook.com/ua-popularity`), store it in
   nixpkgs.
4. Allow passing downloaded packages using environment variable so user
   can override them when needed

bryango commented:
I have taken a look at the https://code.calibre-ebook.com/ua-popularity code path and it seems that one can dump the whole thing as `json` after running `browser_data.get_data()`, and then jump to the logic here:

https://github.com/kovidgoyal/calibre/blob/master/setup/resources.py#L167

wrvsrx commented:
> I have taken a look at the `https://code.calibre-ebook.com/ua-popularity` code path, and it seems that one can dump the whole thing as `json` after running `browser_data.get_data()`, and then jump to the logic here:
> 
> https://github.com/kovidgoyal/calibre/blob/master/setup/resources.py#L167
> 
> If [github search](https://github.com/search?q=repo%3Akovidgoyal%2Fcalibre+browser_data&type=code) is accurate, this is the only place that `browser_data` is ever used.
> 
> In any case, thank you for this epic rewrite!

I think since we just want to patch the downloading behavior, the closer the patch part is logically to the place where the download is called, the better, and the text stored in nixpkgs will be smaller.

bryango commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -28,15 +27,59 @@
+, translations-src ? null
 }:
+let
+  nullOrX = x: y: if x == null then y else x;
```
Nit: instead of `nullOrX`, how about calling this `setDefaultPackage`? Sounds more descriptive.

wrvsrx replied:
Now this function is removed.

bryango commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -51,6 +94,11 @@ stdenv.mkDerivation (finalAttrs: {
+      name = "build-from-git-without-internet";
+      url = "https://github.com/kovidgoyal/calibre/commit/4b7c79f71821e9d13edff5d04249c5decf1093a0.patch";
+      hash = "sha256-1czbc5Z5BsvtyVWnQwFH3Wsn+31RfahjulwtXHJVQKA=";
+    })
```
Is there any chance of upstreaming this to calibre in the future?

wrvsrx replied:
I just made PR to calibre (https://github.com/kovidgoyal/calibre/pull/2193).

bryango commented on pkgs/applications/misc/calibre/fetch-ua.py:
```diff
@@ -0,0 +1,17 @@
+    f.write(content)
+    # to pass nixpkgs newline check
+    f.write('\n')
+subprocess.check_call(['rm', 'result'])
```
Can this be made into a `passthru.updateScript` in the future? See: https://nixos.org/manual/nixpkgs/unstable/#var-passthru-updateScript

wrvsrx replied:
Done.

bryango reviewed (approved):
Approved. I added some nitpicks & questions but they are non-blocking to me. I am not a committer though...

Result of `nixpkgs-review pr 289684` run on `x86_64-linux`
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>calibre</li>
    <li>unbook</li>
  </ul>
</details>

wrvsrx committed:
calibre: refactor function params and add updateScript

eli-schwartz commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -51,6 +90,11 @@ stdenv.mkDerivation (finalAttrs: {
     })
+    (fetchpatch {
+      name = "build-from-git-without-internet";
+      url = "https://github.com/kovidgoyal/calibre/commit/4b7c79f71821e9d13edff5d04249c5decf1093a0.patch";
```
IMO it's quite bad for auditability to list this patch file as inheriting from https://github.com/kovidgoyal/calibre instead of from your fork, since you're relying on github making any commit available from any fork here and as a result it looks superficially to someone who doesn't double check, that the commit was actually upstreamed (which it wasn't).

```suggestion
      url = "https://github.com/wrvsrx/calibre/commit/4b7c79f71821e9d13edff5d04249c5decf1093a0.patch";
```

wrvsrx committed:
calibre: rename author in patch url

Co-authored-by: Eli Schwartz <eschwartz93@gmail.com>

bryango commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -134,20 +179,39 @@ stdenv.mkDerivation (finalAttrs: {
-    export FC_LIB_DIR=${fontconfig.lib}/lib
-    export PODOFO_INC_DIR=${podofo.dev}/include/podofo
-    export PODOFO_LIB_DIR=${podofo.lib}/lib
-    export XDG_DATA_HOME=$out/share
-    export XDG_UTILS_INSTALL_MODE="user"
```
`XDG_*` variables seem to be missing in the new version. Are they not necessary anymore?

eli-schwartz replied:
I rendered this obsolete in https://github.com/kovidgoyal/calibre/pull/988 back in 2019.

wrvsrx replied:
My fault. After some test I found `$out/share/{applications,icons,mime}` would disappear without these environment variables. I have added them back.

eli-schwartz replied:
But why are they needed? calibre will detect that it is run in packaging mode and set it to the correct datadirs without you messing around with INSTALL_MODE...

bryango commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -134,20 +179,39 @@ stdenv.mkDerivation (finalAttrs: {
+    FC_LIB_DIR = "${fontconfig.lib}/lib";
+    PODOFO_INC_DIR = "${podofo.dev}/include/podofo";
+    PODOFO_LIB_DIR = "${podofo.lib}/lib";
+    ISOCODE_ZIP = "${iso-codes-zip}";
+    ISOCODE_VERSION = iso-codes-zip.version;
+    CACERT = "${cacert}/etc/ssl/certs/ca-bundle.crt";
+    TRANSLATIONS = "${translations}";
+    # this is get from https://code.calibre-ebook.com/ua-popularity
+    # It's not reproducible, so we have to pack it in repo
+    # this file can be generated using `python fetch-ua.py`
+    UA_POPULARITY = "${./ua-popularity.txt}";
```
Nit: I feel like it would be better to prefix all these new variables with `CALIBRE_BOOTSTRAP_`, e.g. `CALIBRE_BOOTSTRAP_CACERT`. In particular, `CACERT` and `TRANSLATIONS` are very generic variable names, such that I am afraid that they might get misused by other dependencies.

wrvsrx replied:
Now I use a simpler patch and `CACERT` and `TRANSLATIONS` are no longer needed. `ISOCODE_VERSION` and `ISOCODE_ZIP` is not changed, since it seems calibre itself uses environment variables without prefix, for example, `POPPLER_INC_DIR`.

bryango commented on pkgs/applications/misc/calibre/default.nix:
```diff
@@ -51,6 +90,11 @@ stdenv.mkDerivation (finalAttrs: {
     })
+    (fetchpatch {
+      name = "build-from-git-without-internet";
+      url = "https://github.com/kovidgoyal/calibre/commit/4b7c79f71821e9d13edff5d04249c5decf1093a0.patch";
```
Since this is unofficial, I believe it would be easier for people to maintain the patch by putting it in nixpkgs directory. For example the change of variable names `CACERT -> CALIBRE_BOOTSTRAP_CACERT` can be more easily accomplished this way. What do you think?

wrvsrx replied:
You are right. Done.

bryango reviewed (approved):
Still approved, but I added some more comments, all of them are optional, non-blocking suggestions. Sorry for the never-ending comments haha...

Finally, I think now you are an expert on this package now; would you consider adding yourself to `meta.maintainers`? In the future, maintainers will have the right to commit to their own packages: https://discourse.nixos.org/t/nixpkgs-merge-bot-testing-and-plan/39824

wrvsrx committed:
calibre: use a simpler patch to build from git source

Since upstream refused the PR which allow offline bootstrap from git
source, we don't need to try to make patch compatible with online
bootstrap. This commit uses a simpler and more robost patch.

This commit also adds `XDG` environment variables back during build.

wrvsrx committed:
calibre: add the non-offical patch to nixpkgs

eli-schwartz reviewed (changes requested):
My overall opinion is as mentioned in the upstream PR -- I do not believe it makes sense to build from the git source, as the reasoning for doing so is not at all clear.

From this nixpkgs PR:
> However, the tarball contains lots of generated code such as javascript generated from python, python generated from pyqt... It also contained generated man pages.

... why is it bad to have "lots of generated code"? Do you use git sources instead of tarballs for autotools software and recreate the ./configure script, too? What are the *precise* reasons for avoiding "generated code"?

From the upstream ticket:
> Because the release tarball contains some generated js code and pyqt code, patching python code in tarball cannot modify the that generated code.
> 
> For example I want to apply https://github.com/kovidgoyal/calibre/commit/14caed3a6f0f57e172d1eb407703f0e4cc179e11 to fix https://www.mobileread.com/forums/showthread.php?t=359082 before next release.

But this is not a reason to *avoid* generated code. Simply use some generated code, and some patched code. Simplicity itself.

I'm not a NixOS user, but it seems to me like NixOS very often tries to make things intentionally hard for NixOS, and it's not clear *why*.

wrvsrx commented:
> ... why is it bad to have "lots of generated code"?

Why not build from rawer source?

> But this is not a reason to avoid generated code. Simply use some generated code, and some patched code. Simplicity itself.

Suppose we use tarball:

1. If we don't regenerate the code when packaging, then user patching will be more complicated than building from git source. If one want to apply a patch, they need to first understand how the package is built, secondly patch code and then change corresponding buildPhase to regenerate the code. It building from git source, they just need to apply that patch then build system will help them do following thing. It's worthy to note that creating a small patch to generated code is hard since the generated code is easy to differ a lot from that in tarball when library versions are different, even no code change happens in git source. I have given an example in upstream PR.
1. If we regenerate the code when packaging, we need to findout all code generating stages and make sure all old generated code is covered. Why not just use a clean environment? Building from git source simply predownloads needed resources in a reproducible manner, runs bootstrap, man_pages and install.

bryango commented:
> > Because the release tarball contains some generated js code and pyqt code, patching python code in tarball cannot modify the that generated code.
> > For example I want to apply [kovidgoyal/calibre@14caed3](https://github.com/kovidgoyal/calibre/commit/14caed3a6f0f57e172d1eb407703f0e4cc179e11) to fix [mobileread.com/forums/showthread.php?t=359082](https://www.mobileread.com/forums/showthread.php?t=359082) before next release.
> 
> But this is not a reason to _avoid_ generated code. Simply use some generated code, and some patched code. Simplicity itself.
> 
> I'm not a NixOS user, but it seems to me like NixOS very often tries to make things intentionally hard for NixOS, and it's not clear _why_.

The whole point is:
- As long as we need to patch the git source, we need to re-run the bootstrap _anyway_.
- The bootstrap is highly non-reproducible as it pulls un-versioned stuff from the internet.
- In nix, random pulls from the internet have to be isolated out and pinned to be reproducible. 

> NixOS very often tries to make things intentionally hard for NixOS

I do agree with this sometimes, that's why I use nix as a package manager on Arch (not NixOS itself) :laughing: in this case, yes, nix intentionally make things harder, but it does so for good reasons: reproducibility.

bryango reviewed (approved):
The patch is looking a lot more manageable now! Approved with no more comments.

eli-schwartz commented:
> If we don't regenerate the code when packaging, then user patching will be more complicated than building from git source.

You're not even using the PGP signed sources followed by regenerating the code. This is bad.

> If one want to apply a patch, they need to first understand how the package is built, secondly patch code and then change corresponding buildPhase to regenerate the code. It building from git source, they just need to apply that patch then build system will help them do following thing.

Yes, this is the reason why distros have package maintainers... also the reason why testing your changes exists...

The problem here is that if you regenerate tons of stuff you don't need, then you add pointless, frivolous, wasteful dependencies that bother end users.
- Dependencies on sphinx, for example.
- You also *add* additional unwarranted problems for assets that you have no valid use case for modifying, and then have to apply patches to do that offline:
  - isocodes
  - translations
  - cacerts
  - user agents
 
This is just throwing the baby out with the bathwater because it's not worth your time and effort as a maintainer of a NixOS package to understand the software and keep on top of what is needed to effectively maintain it while simultaneously posing the least burden upon users.

> It's worthy to note that creating a small patch to generated code is hard since the generated code is easy to differ a lot from that in tarball when library versions are different, even no code change happens in git source. I have given an example in upstream PR.

I keep on saying this and maybe my words will eventually be understood: I am not, nor have I ever, recommended adding patches for generated code. You invented that idea just so you could explain why it's bad, which is otherwise known as a "strawman argument".

Regenerate the code you actually need to regenerate. That's it.

eli-schwartz commented:
My overall opinion is 

> The whole point is:
> 
>   * As long as we need to patch the git source, we need to re-run the bootstrap _anyway_.
> 
>   * The bootstrap is highly non-reproducible as it pulls un-versioned stuff from the internet.
> 
>   * In nix, random pulls from the internet have to be isolated out and pinned to be reproducible.

As a major contributor to calibre who has expended vast effort into improving calibre's build system *specifically*, and done so because I want to make it be better able to be packaged in Linux distros:

You are incorrect on all points.

For your first point, you're incorrect, you don't have to run bootstrap ever. I attempted to inform @wrvsrx in the linked upstream PR, but I was ignored and instead this patch is applied to NixOS which modifies the calibre build system.

For your second point, the bootstrap is NOT highly non-reproducible, since I extensively reworked the upstream build system to make it merely "minimally non-reproducible".

For your third point, it is a red herring because nix doesn't have to do random pulls from the internet if you use the non-git source.

wrvsrx commented:
> Dependencies on sphinx, for example.

It a program need use sphinx to generate man page, why is it wasteful?

> You also add additional unwarranted problems for assets that you have no valid use case for modifying, and then have to apply patches to do that offline:
> - isocodes
> - translations
> - cacerts
> - user agents

It is normal to use the system's own package to replace the package of the software bundle when packaging the distribution version, isocodes and cacerts are from nixpkgs itself. `user-agents-data.json` is a file changing every minute, why do you think it will burden user. `translations` has the same git rev as that used in calibre 7.5.1.

> > It's worthy to note that creating a small patch to generated code is hard since the generated code is easy to differ a lot from that in tarball when library versions are different, even no code change happens in git source. I have given an example in upstream PR.
> 
> I keep on saying this and maybe my words will eventually be understood: I am not, nor have I ever, recommended adding patches for generated code. You invented that idea just so you could explain why it's bad, which is otherwise known as a "strawman argument".

Here I will give explaination on two patch methods. One is patch first and then regenerate the code as you mentioned before, and the other is the js code generated by patch mentioned by the caliber author. I did not invent this latter idea.

bryango commented:
> You are incorrect on all points.
> 
> For your first point, you're incorrect, you don't have to run bootstrap ever.

Ah I see, thank you for that information; then the following two points are moot and maybe wrvsrx can simplify the patch accordingly. However,

> For your first point, you're incorrect, you don't have to run bootstrap ever. I attempted to inform @wrvsrx in the linked upstream PR, but I was ignored and instead this patch is applied to NixOS which modifies the calibre build system.
> 
> For your second point, the bootstrap is NOT highly non-reproducible, since I extensively reworked the upstream build system to make it merely "minimally non-reproducible".
> 
> For your third point, it is a red herring because nix doesn't have to do random pulls from the internet if you use the non-git source.
> 
> > in this case, yes, nix intentionally make things harder, but it does so for good reasons: reproducibility.
> 
> No, the good reason is a misconception.

Is this _documented_ somewhere? Thank you for working on calibre for all of us for all these years, including the epic python3 migration. But please understand that I am a mere user and might have accidentally not followed a comment under a closed PR. In any case, thank you for clearing up the puzzles, at least for me. Hopefully wrvsrx can simplify the patch accordingly.

eli-schwartz commented:
> It a program need use sphinx to generate man page, why is it wasteful?

It's wasteful if the source tarball already has the man page, and users would otherwise not need to download, build, and install the sphinx documentation generator.

> It is normal to use the system's own package to replace the package of the software bundle when packaging the distribution version, isocodes and cacerts are from nixpkgs itself. `user-agents-data.json` is a file changing every minute, why do you think it will burden user. `translations` has the same git rev as that used in calibre 7.5.1.

Why futz around with the translations at all?

If you are so concerned about using the system's own package for isocodes and cacerts, I would think you'd be significantly more concerned with the several other resources that are not using system copies (which I designed the calibre build system to utilize system copies for)?

Yes, I think it burdens the user to mess with user-agents-data instead of using the prebundled one. Moreover, however, I think it burdens NixOS to apply patches to calibre to download a new version of that file "every minute".

Note also that cacerts is kind of a pointless waste of time IMO as it is only used on macOS, and only then to embed it into bypy/macos/launcher.c (used for creating the .dmg installer).



> Here I will give explaination on two patch methods. One is patch first and then regenerate the code as you mentioned before, and the other is the js code generated by patch mentioned by the caliber author. I did not invent this latter idea.

Perhaps I misinterpreted this:

> Because the release tarball contains some generated js code and pyqt code, patching python code in tarball cannot modify the that generated code.

It looked to me like you were suggesting patching the generated code but saying that you find it problematic to do so, and the calibre author was responding by saying "actually if you really want to you can, the diff isn't that big".

wrvsrx commented:
> Yes, I think it burdens the user to mess with user-agents-data instead of using the prebundled one. Moreover, however, I think it burdens NixOS to apply patches to calibre to download a new version of that file "every minute".

It's totally wrong. This file changes every time calibre got bootstrap, the patch fix it.

> For your second point, the bootstrap is NOT highly non-reproducible, since I extensively reworked the upstream build system to make it merely "minimally non-reproducible".

It's still non-reproducible. If you bootstrap calibre twice, the checksum of tarball is different.

pSub committed:
Merge branch 'master' into build-calibre-from-source

wrvsrx commented:
I'll open a new PR after calibre with https://github.com/kovidgoyal/calibre/pull/2309 is released.

After https://github.com/kovidgoyal/calibre/pull/2309, we can build calibre from git source without patch it.

eli-schwartz commented:
@wrvsrx I still do not understand why you insist to build calibre from git sources instead of using a release tarball.

kirillrdy commented:
@eli-schwartz https://discourse.nixos.org/t/when-to-use-precompiled-binaries-in-a-nixpkgs-derivation/27087/2

eli-schwartz commented:
It's not a precompiled binary. It is:
- pre-rendered sphinx-generated manpages
- pre-rendered json files containing a list of user agents, vs checking a prerendered json file into nixpkgs.git
- byte-identical copies of some assets included in the release tarball vs downloaded via their own fetchFromGitHub and `shutil.copy`d in


Please.

Explain to me what your ideological purge of pre-rendered manpages is about.

kirillrdy commented:
Apologies. My brain parsed "release tarball" as recompiled binary.

Some points may still apply eg using same list of user agents across nixpkgs but I haven't had a closer look at this specific case

wrvsrx commented:
> @wrvsrx I still do not understand why you insist to build calibre from git sources instead of using a release tarball.

1. User friendness. [Overriding](https://ryantm.github.io/nixpkgs/using/overrides/) is a common pattern in nixpkgs, this PR (and the new PR I will make) make it easier to cherry-pick other patches for users since they just need to apply the patch without changing the build phase and build inputs.
2. Safety. Think about [story of zk](https://news.ycombinator.com/item?id=40017310).
3. Reproducibility.

Some responses to your previous question:

1. 
> Dependencies on sphinx, for example.

In nix `build time dependency` is different from `run time dependency`, sphinx is a build time dependency rather than a run time dependency. Users will only download calibre and its run time dependencies from nixpkgs' binary cache.

2. 
> You also add additional unwarranted problems for assets that you have no valid use case for modifying, and then have to apply patches to do that offline:
>   isocodes
>   translations
>   cacerts
>   user agents

Upstream has accepted my PR, so patch is no longer needed.

---

I was wondering if you could explain why you think building from tarball is better? @eli-schwartz

wrvsrx commented:
@Schweber You could just override `speechd` to `pkgs.empty`, it is optional.

wrvsrx commented:
@Schweber 

You can just override calibre itself, for example:

```nix
{ pkgs, ... }:
{
  environment.systemPackages = [ ( pkgs.calibre.override { speechd = pkgs.empty; } ) ];
}
```

eli-schwartz commented:
> > @wrvsrx I still do not understand why you insist to build calibre from git sources instead of using a release tarball.
> 
> 1. User friendness. [Overriding](https://ryantm.github.io/nixpkgs/using/overrides/) is a common pattern in nixpkgs, this PR (and the new PR I will make) make it easier to cherry-pick other patches for users since they just need to apply the patch without changing the build phase and build inputs.

- I am informing you point blank. Anyone who is using an override to apply a patch to the translations must be prevented from doing so, because it's insane.
- Everything else is either already part of the default build rules, or covered by the fact that you SHOULD be un-vendoring multiple things via a separate runtime dependency, and which I have already told you NOT to download as a fetchFromGitHub. You have ignored me.

In short, you are the enemy of user-friendliness here, not me.

> 2. Safety. Think about [story of zk](https://news.ycombinator.com/item?id=40017310).

This is retroactive justification and not your original (lack of a) reason.

I'm also exceedingly skeptical that it makes sense as a rationale. Leaving aside the misspelling of the name of the thing in question, the xz hack was a story about social engineering, not reproducibility. In fact, the payload was checked into git and present in a git clone.

There were other changes in xz that were part of e.g. the CMakeLists.txt and didn't even require a generated autoconf script.

...

If you want to verify reproducibility, how about you reproducibly generate the source tarball, rather than use source code from git that requires additional dependencies and is only as reproducible as the giant binary blobs you vendor into nixpkgs instead?

Reproducing the tarball would also have the benefit of actually being useful to people outside of nixpkgs.

Note that you can prove most things are not a vulnerability either way. The text in a manpage isn't an attack vector. json lists of user agents aren't an attack vector as long as the source code doesn't try to `exec()` the contents. Many resources can be rm -rf'ed from the distribution tarball and regenerated the same way you would do so from a git clone. And in fact you have to regenerate it in order to correctly devendor assets that I described above should be devendored.


> 3. Reproducibility.

It is infinitely more reproducible to use a pregenerated manpage that never changes, than to generate one yourself that changes depending on the version of sphinx you have.

> 
> Some responses to your previous question:
> 
> 
> > Dependencies on sphinx, for example.
> 
> In nix `build time dependency` is different from `run time dependency`, sphinx is a build time dependency rather than a run time dependency. Users will only download calibre and its run time dependencies from nixpkgs' binary cache.

Thanks, I know. Every other linux distro has the exact same concept.

Do nix users never build packages themselves? But I thought your whole reason for doing this is because users might apply an override to the recipe!

So the target userbase for whom you want to away from PGP-signed tarballs to a git clone, is... people who are affected by needing to download build time dependencies.

> > You also add additional unwarranted problems for assets that you have no valid use case for modifying, and then have to apply patches to do that offline:
> > isocodes
> > translations
> > cacerts
> > user agents
> 
> Upstream has accepted my PR, so patch is no longer needed.

I'm not particularly worried about the patch specifically, so much as I am about the fact that user agents requires checking json blobs into nixpkgs.git, and translations has never required a patch (you just don't understand the calibre build system) but also requires build time dependencies such as sphinx, which is bad for users applying overrides.
